README.md
# SvelteKit CAS authentication

A set of functions to ease usage of a CAS/SSO in SvelteKit

![Github CI](https://github.com/macfja/sveltekit-cas/workflows/Quality%20checks/badge.svg)
![GitHub Repo stars](https://img.shields.io/github/stars/macfja/sveltekit-cas?style=social)
![NPM bundle size](https://img.shields.io/bundlephobia/minzip/@macfja/sveltekit-cas)
![Download per week](https://img.shields.io/npm/dw/@macfja/sveltekit-cas)
![License](https://img.shields.io/npm/l/@macfja/sveltekit-cas)
![NPM version](https://img.shields.io/npm/v/@macfja/sveltekit-cas)
![Snyk Vulnerabilities for npm package](https://img.shields.io/snyk/vulnerabilities/npm/@macfja/sveltekit-cas)

## Installation

```shell
npm install --save @macfja/sveltekit-cas
```

## Usage

Protect all pages that start with `/profile/` and only allow user to go on his own page (`/profile/my-cool-username`)

```typescript
// src/hooks.server.ts
import { sessionHook } from "@macfja/sveltekit-session"
import { casHandler } from "@macfja/sveltekit-cas"
import type { Handle } from "@sveltejs/kit"
import { sequence } from "@sveltejs/kit/hooks"

export const handle: Handle = sequence(
	sessionHook(),
	casHandler(
		"http://0.0.0.0:8080",
		2,
		(event) => event.url.pathname.startsWith("/profile/"),
		(event, user) => {
			const regexp = event.url.pathname.match(/\/profile\/(\w+)/)
			return user !== regexp[1]
		}
	)
)
```

Protect endpoint, so only connected user can access it

```typescript
// src/routes/api/user/server.ts
import { error } from "@sveltejs/kit"
import { getUsername } from "@macfja/sveltekit-cas"
import type { RequestHandler } from "./$types"

export const POST: RequestHandler = async (event) => {
	if (getUsername(event) !== "admin") {
		throw error(403)
	}

	// ... Do operation that only the user `admin` can do
}

export const GET: RequestHandler = async (event) => {
	if (getUsername(event) === undefined) {
		throw error(401)
	}

	// ... Do operation that only connected user can do
}
```

## Configuration

The `casHandler` function take 4 parameters to change its behavior:

- `casRoot`: The root URL to the CAS server
- `casVersion`: The version of the CAS server (supported version: `1`, `2`, `3`)
- `authRequired` _(optional)_: A function to indicate if a request should have an authenticated user
- `rejectAccess` _(optional)_: A function to indicate if a particular authenticated user if allowed to do a request

## Contributing

Contributions are welcome. Please open up an issue or create PR if you would like to help out.

Read more in the [Contributing file](CONTRIBUTING.md)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.2]

### Fixed

- Change permanente redirection to temporary redirection
- Remove namespace in XML received from the CAS ([Issue#1])
- Handle response that are not an XML from the CAS ([Issue#1])

### Added

- Add a new parameter to handle trailing slash in URL ([Issue#1])

## [2.0.1]

### Fixed

- Fix import failure with CommonJS ([Issue#1])

## [2.0.0]

### Changed

- Add compatibility to all SvelteKit version `1.0.0-next.492` and later
- Use [@macfja/sveltekit-session](https://www.npmjs.com/package/@macfja/sveltekit-session) to data persistence

## [1.0.1]

### Fixed

- Import of `njwt` not compatible
- Session cookie not HTTP only and not Secure

## [1.0.0]

First version

[unreleased]: https://github.com/MacFJA/sveltekit-cas/compare/2.0.2...HEAD
[2.0.2]: https://github.com/MacFJA/sveltekit-cas/releases/tag/2.0.2
[2.0.1]: https://github.com/MacFJA/sveltekit-cas/releases/tag/2.0.1
[2.0.0]: https://github.com/MacFJA/sveltekit-cas/releases/tag/2.0.0
[1.0.1]: https://github.com/MacFJA/sveltekit-cas/releases/tag/1.0.1
[1.0.0]: https://github.com/MacFJA/sveltekit-cas/releases/tag/1.0.0
[issue#1]: https://github.com/MacFJA/sveltekit-cas/issues/1


src/cas.ts
import env from "env-var"
import { CASClientV2 } from "logical-cas-client"
import njwt from "njwt"
const create = njwt.create,
	verify = njwt.verify

/**
 * @internal
 * @param {string} username
 */
const createJwt = (username: string) =>
	create(
		{
			iss: env.get("JWT_ISS").default("sveltekit-cas").asString(),
			sub: "cas/" + username,
			scope: "user"
		},
		env.get("JWT_SECRET").default("changeme").asString()
	)

/**
 * @internal
 * @param {string} jwtString
 */
const readJwt = (jwtString: string) => {
	return verify(jwtString, env.get("JWT_SECRET").default("changeme").asString())
}
/**
 * @internal
 * @param jwtString
 */
export const getJwtUser = (jwtString: string): string | null => {
	try {
		const jwt = readJwt(jwtString)
		return jwt.body.toJSON()?.sub?.toString().substring(4)
	} catch (e) {
		return null
	}
}

/**
 * @internal
 * @param {string} path
 * @param {(user: string) => void} onSuccess
 * @param {(error, req) => void} onFailure
 */
const createCas = (path: string, onSuccess, onFailure): CASClientV2 => {
	return new CASClientV2(
		{
			host: env.get("PUBLIC_HOST").required().asString(),
			secure: true,
			port: env.get("PUBLIC_PORT").default(443).asPortNumber(),
			server: {
				host: env.get("CAS_HOST").required().asString(),
				port: env.get("CAS_PORT").default(443).asPortNumber(),
				version: env.get("CAS_VERSION").default("3.0").asString(),
				secure: true
			},
			endpoints: {
				ticketVerificationPath: path
			}
		},
		(req, res, user) => onSuccess(user),
		(req, res, error) => onFailure(error, req)
	)
}

/**
 * @internal
 * @param session
 * @param path
 * @param ssoTicket
 */
export const getCasUser = async (
	session: string | null,
	path: string,
	ssoTicket?: string
): Promise<{ session: string; user?: string; redirect?: string }> => {
	if (session) {
		const jwtUser = getJwtUser(session)
		if (jwtUser) {
			return {
				session,
				user: getJwtUser(session)
			}
		}
	}

	let ssoUser
	let redirect
	const cas = createCas(
		path,
		(user) => (ssoUser = user),
		() => (redirect = "/")
	)

	if (ssoTicket) {
		await cas.verifyTicket(
			{
				path: path,
				query: { ticket: ssoTicket }
			},
			{}
		)
		const jwt = createJwt(ssoUser)
		return {
			redirect: path,
			user: ssoUser,
			session: jwt.compact()
		}
	}

	await cas.redirectToCASLogin(
		{},
		{
			redirect: (url) => {
				redirect = url
			}
		}
	)
	return {
		redirect,
		session
	}
}


src/hooks.ts
import type { ServerRequest, ServerResponse } from "@sveltejs/kit/types/hooks"
import env from "env-var"
import Cookie from "universal-cookie"
import { getCasUser, getJwtUser } from "./cas"

const cookieName = env.get("CAS_SESSION_COOKIE").default("session").asString()

/**
 * Hooks handler for CAS.
 *
 * This function is intended to by use in the `handle` function of your `src/hooks.js`.
 *
 * Like for example:
 * ```javascript
 * // src/hooks.js
 * import { casHandler } from "@macfja/sveltekit-cas"
 *
 * export async function handle({ request, resolve }) {
 * 	return (
 * 		(await casHandler(
 * 			// All URL prefixed by `/profile/` will be behind the CAS
 * 			(request) => request.path.startsWith("/profile/"),
 * 			// Deny access if the username from the CAS is not the same as the one in the URL
 * 			(request, user) => {
 * 				const regexp = request.path.match(/\/profile\/(\w+)/)
 * 				return user !== regexp[1]
 * 			},
 * 			request
 * 		)) || resolve(request)
 * 	)
 * }
 * ```
 *
 * @param {(request: ServerRequest) => boolean} filterOn
 *    Function to determine if a request should be authenticated.
 *    If you want all requests to be authenticated, then always return true: `() => true`
 * @param {(request: ServerRequest, user: string) => boolean} rejectOn
 *    Function to reject an authenticated user from a page or not.
 *    If you only want a user to be connected, then always return false: `() => false`
 * @param {ServerRequest<Record<string, any>, Body>} request
 *    The server request
 *
 * @return {Promise<ServerResponse | null>}
 *    Return a promise that resolve to:
 *
 *    - `null` if the access is **GRANTED**
 *    - a _redirection_ (`ServerResponse`) if the user need to go to the CAS login page
 *    - a _403_ (`ServerResponse`) if the access is **DENIED**
 */
export async function casHandler(
	filterOn: (request: ServerRequest) => boolean,
	rejectOn: (request: ServerRequest, user: string) => boolean,
	request: ServerRequest<Record<string, any>, Body>
): Promise<ServerResponse | null> {
	const cookies = new Cookie(request.headers.cookie)

	if (!filterOn(request)) {
		return null
	}

	const data = await getCasUser(cookies.get(cookieName), request.path, request.query.get("ticket"))
	if (data.redirect) {
		return {
			status: 302,
			headers: {
				location: data.redirect,
				"set-cookie": cookieName + "=" + (data.session ?? "") + ";Secure;HttpOnly"
			}
		}
	}
	if (rejectOn(request, data.user)) {
		return {
			status: 403,
			headers: {}
		}
	}
	return null
}

/**
 * Get token used to know if the user is connected.
 * This the value that you will need to communicate to your endpoint to validate that the request is made by an authenticated user.
 *
 * ```javascript
 * // src/hooks.js
 * import { getSessionToken } from "@macfja/sveltekit-cas"
 * export function getSession(request) {
 * 	return {
 * 		...getSessionToken(request)
 * 	}
 * }
 * ```
 *
 * @param {ServerRequest} request The request of the current HTTP call
 * @return {{ token: string | null }} Return an object that contain one key named `token`, with the value to the token string or `null` (if the user is not authenticated)
 */
export function getSessionToken(request: ServerRequest): { token: string | null } {
	const cookies = new Cookie(request.headers.cookie)
	return {
		token: cookies.get(cookieName) ?? null
	}
}
/**
 * Get the user name of the connected user.
 *
 * ```javascript
 * // src/hooks.js
 * import { getSessionUser } from "@macfja/sveltekit-cas"
 * export function getSession(request) {
 * 	return {
 * 		...getSessionUser(request)
 * 	}
 * }
 * ```
 *
 * @param {ServerRequest} request The request of the current HTTP call
 * @return {{ user: string | null }} Return an object that contain one key named `user`, with the value to the username or `null` (if the user is not authenticated)
 */
export function getSessionUser(request: ServerRequest): { user: string | null } {
	const cookies = new Cookie(request.headers.cookie)
	return {
		user: getJwtUser(cookies.get(cookieName))
	}
}


src/endpoint.ts
import type { EndpointOutput } from "@sveltejs/kit"
import { getJwtUser } from "./cas"

/**
 * Get the username from the token.
 * @internal
 *
 * @param {string | null} token
 * @return {string | null} The username or `null`
 */
function getUsername(token: string | null): string | null {
	if (token === null) {
		return null
	}

	return getJwtUser(token)
}

/**
 * Validate is the token is valid:
 *  - Check if the token is not expired
 *  - Check if the token have been created by the application
 *  - Check if the token contain an username
 *
 * ```javascript
 * // src/routes/api/user.js
 * import { validate } from "@macfja/sveltekit-cas"
 *
 * export async function get({ headers }: ServerRequest): Promise<EndpointOutput> {
 * 	const token = headers.token ?? null
 * 	const access = validate(token)
 *
 * 	if (access !== null) {
 * 		return access
 * 	}
 *
 * 	// ... Do operation that only authenticated user can do
 * }
 * ```
 * @param {string | null} token
 *    The token to validate
 *
 * @return {EndpointOutput | null}
 *    Return
 *
 *    - `null` if the token is valid
 *    - _401_ (`EndpointOutput`) if the token is **INVALID**
 */
export function validate(token: string | null): EndpointOutput | null {
	const userName = getUsername(token)

	if (userName === null) {
		return {
			status: 401
		}
	}

	return null
}

/**
 * Validate is the token is valid and match the requested user:
 *  - Check if the token is not expired
 *  - Check if the token have been created by the application
 *  - Check if the token contain an username
 *  - Check if the username in the token is the same as the requested username
 *
 * ```javascript
 * // src/routes/api/user.js
 * import { validateUser } from "@macfja/sveltekit-cas"
 *
 * export async function post({ headers, body }: ServerRequest): Promise<EndpointOutput> {
 * 	const token = headers.token ?? null
 * 	const username = body.get('username')
 * 	const access = validateUser(token, username ?? '')
 *
 * 	if (access !== null) {
 * 		return access
 * 	}
 *
 * 	// ... Do operation that only the user withe username `username` can do
 * }
 * ```
 * @param {string | null} token
 *    The token to validate
 * @param {string} user
 *    The username to validate against
 *
 * @return {EndpointOutput | null}
 *    Return
 *
 *    - `null` if the token is valid and teh username are the same
 *    - _401_ (`EndpointOutput`) if the token is **INVALID**
 *    - _403_ (`EndpointOutput`) if the username are **DIFFERENT**
 */
export function validateUser(token: string | null, user: string): EndpointOutput | null {
	const userName = getUsername(token)

	if (userName === null) {
		return {
			status: 401
		}
	}

	if (userName !== user) {
		return {
			status: 403
		}
	}

	return null
}


src/index.ts
import type { Handle, RequestEvent } from "@sveltejs/kit"
import type { SessionLocals } from "@macfja/sveltekit-session"
import { js2xml, xml2js } from "xml-js"

type ServiceValidation = {
	serviceResponse: {
		authenticationFailure?: ServiceValidationFailure
		authenticationSuccess?: ServiceValidationSuccess
	}
}
type ServiceValidationFailure = {
	_attributes: {
		code: string
	}
	_text: string
}
type ServiceValidationSuccess = {
	user: {
		_text: string
	}
	attributes: Record<string, string>
}
export type TrailingSlashFunction = (
	route: Pick<RequestEvent, "route">["route"]
) => "never" | "always" | "ignore"
export type CasSession = {
	/** The username of the current authenticated user. */
	user: string
	/** Any additional data on the user return by the CAS server. */
	attributes: Record<string, unknown>
}

const getCasHelper = (
	root: string,
	version: 1 | 2 | 3,
	event: RequestEvent,
	trailing: TrailingSlashFunction
) => {
	return {
		getValidationUrl(ticket: string): string {
			let validationPath = ""
			switch (version) {
				case 1:
					validationPath = "validate"
					break
				case 2:
					validationPath = "serviceValidate"
					break
				case 3:
					validationPath = "p3/serviceValidate"
			}
			return `${root}/cas/${validationPath}?service=${encodeURI(
				this.sanitizedService()
			)}&ticket=${ticket}`
		},
		getLoginUrl(): string {
			return root + "/cas/login?service=" + encodeURI(this.sanitizedService())
		},
		sanitizedService(): string {
			const newUrl = new URL(event.url)
			if (newUrl.pathname.endsWith("/__data.json")) {
				newUrl.pathname = newUrl.pathname.substring(
					0,
					newUrl.pathname.length - "__data.json".length
				)
			}
			newUrl.searchParams.delete("ticket")
			newUrl.searchParams.delete("x-sveltekit-invalidated")
			const urlString = newUrl.toString()

			switch (trailing(event.route)) {
				case "ignore":
					return urlString
				case "never":
					return urlString.substring(0, urlString.length - (urlString.endsWith("/") ? 1 : 0))
				case "always":
					return urlString + (urlString.endsWith("/") ? "" : "/")
			}
		}
	}
}
function getTicket(event: RequestEvent): string | undefined {
	if (!event.url.searchParams.has("ticket")) return undefined
	const ticket = event.url.searchParams.get("ticket") as string
	return ticket.startsWith("ST-") ? ticket : undefined
}
function redirect(url: string): Response {
	return new Response(null, {
		status: 302,
		headers: {
			location: url
		}
	})
}

/**
 * SvelteKit server Hook Handler.
 * @example
 * ```typescript
 * // src/hooks.server.ts
 * import { sessionHook } from "@macfja/sveltekit-session";
 * import { casHandler } from "@macfja/sveltekit-cas"
 * import type { Handle } from "@sveltejs/kit"
 * import { sequence } from "@sveltejs/kit/hooks"
 *
 * export const handle: Handle = sequence(sessionHook(), casHandler(
 * 	 'http://0.0.0.0:8080',
 * 	 2,
 * 	 (event) => event.url.pathname.startsWith('/sverdle'),
 * 	 () => false,
 * 	 (route) => route.id === '/blog/[slug]' ? 'never' : 'ignore'
 * ))
 * ```
 * @param casRoot
 * 		The Root url of the CAS server
 * @param casVersion
 * 		The version of the CAS server
 * @param authRequired
 * 		Define if a request (first parameter) should be behind the CAS (return true if the request need to be authenticated).
 * 		By default, all request need authentication.
 * @param rejectAccess
 * 		Define if for a request (first parameter) the authentication of a user (second parameter) is allowed (return true if the user is not allowed).
 * 		By default, there are no restriction on the authenticated user.
 * @param trailingSlash
 * 		Define if the trailing slash of a route should be added/removed or left untouched
 * 		By default, the trailing slash is left untouched
 * @return {Handle}
 */
export function casHandler(
	casRoot: string,
	casVersion: 1 | 2 | 3,
	authRequired: (event: RequestEvent) => boolean = () => true,
	rejectAccess: (event: RequestEvent, authenticationInformation: CasSession) => boolean = () =>
		false,
	trailingSlash: TrailingSlashFunction = () => "ignore"
): Handle {
	return function (input: Parameters<Handle>[0]): ReturnType<Handle> {
		if (!authRequired(input.event)) {
			return input.resolve(input.event)
		}

		if ((input.event.locals as SessionLocals).session.cas) {
			if (
				rejectAccess(input.event, (input.event.locals as SessionLocals).session.cas as CasSession)
			) {
				return new Response(null, {
					status: 403
				})
			}
			return input.resolve(input.event)
		}

		switch (casVersion) {
			case 1:
				return cas1Hook(casRoot, input.event, trailingSlash)
			case 2:
				return cas23Hook(casRoot, 2, input.event, trailingSlash)
			case 3:
			default:
				return cas23Hook(casRoot, 3, input.event, trailingSlash)
		}
	}
}

async function cas1Hook(
	casRoot: string,
	event: RequestEvent,
	trailing: TrailingSlashFunction
): Promise<Response> {
	const casHelper = getCasHelper(casRoot, 1, event, trailing)
	let ticket: undefined | string
	if ((ticket = getTicket(event)) !== undefined) {
		const validation: Array<string> = await event
			.fetch(casHelper.getValidationUrl(ticket))
			.then((response) => response.text())
			.then((response) => response.split("\n"))

		if (validation.length === 2) {
			if (validation[0] === "no") {
				return new Response("Authentication failure", { status: 403 })
			}
		}
		if (validation.length === 3) {
			if (validation[0] === "yes") {
				;(event.locals as SessionLocals).session.cas = {
					user: validation[1],
					attributes: {}
				}
				return redirect(casHelper.sanitizedService())
			}
		}
		return new Response("Unknown CAS validation response", { status: 500 })
	}

	return redirect(casHelper.getLoginUrl())
}
async function cas23Hook(
	casRoot: string,
	version: 2 | 3,
	event: RequestEvent,
	trailing: TrailingSlashFunction
): Promise<Response> {
	const casHelper = getCasHelper(casRoot, version, event, trailing)
	let ticket: undefined | string
	// Forcefully remove namespace in XML node element
	const namespaceRemover = (name: string) => (/:/.test(name) ? name.split(":")[1] : name)
	if ((ticket = getTicket(event)) !== undefined) {
		const validation: ServiceValidation = await event
			.fetch(casHelper.getValidationUrl(ticket))
			.then((response) => response.text())
			.then(
				(response) =>
					xml2js(response, { compact: true, elementNameFn: namespaceRemover }) as ServiceValidation
			)
			.catch((reason) => {
				console.error(
					"An error occur while validating the CAS ticket: " + reason,
					casHelper.getValidationUrl(ticket as string)
				)
				return {} as ServiceValidation
			})

		if (!Object.keys(validation).includes("serviceResponse")) {
			console.error(
				"Unknown CAS validation response",
				casHelper.getValidationUrl(ticket),
				js2xml(validation)
			)
			return new Response("Unknown CAS validation response", { status: 500 })
		}
		if (Object.keys(validation.serviceResponse).includes("authenticationFailure")) {
			return new Response(
				`Authentication failure: ${
					(validation.serviceResponse.authenticationFailure as ServiceValidationFailure)._text
				}`,
				{ status: 403 }
			)
		}
		if (Object.keys(validation.serviceResponse).includes("authenticationSuccess")) {
			;(event.locals as SessionLocals).session.cas = {
				user: (validation.serviceResponse.authenticationSuccess as ServiceValidationSuccess).user
					._text,
				attributes: (validation.serviceResponse.authenticationSuccess as ServiceValidationSuccess)
					.attributes
			}
			return redirect(casHelper.sanitizedService())
		}
		return new Response("Unknown CAS validation response", { status: 500 })
	}

	return redirect(casHelper.getLoginUrl())
}

/**
 * Get the username of the authenticated user
 * @param event The current RequestEvent
 * @return `undefined` if there are not connected user
 */
export function getUsername(event: RequestEvent): string | undefined {
	return ((event.locals as SessionLocals).session?.cas as CasSession)?.user
}
